extensions/sql/processors/ExecuteSQL.cpp

```diff
@@ -0,0 +1,123 @@
+#include <codecvt>
+
+#include <soci/soci.h>
+
+#include "io/DataStream.h"
+#include "core/ProcessContext.h"
+#include "core/ProcessSession.h"
+#include "Exception.h"
+#include "utils/OsUtils.h"
+#include "data/DatabaseConnectors.h"
+#include "data/JSONSQLWriter.h"
+#include "data/SQLRowsetProcessor.h"
+#include "data/WriteCallback.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+const std::string ExecuteSQL::ProcessorName("ExecuteSQL");
+
+const core::Property ExecuteSQL::s_sqlSelectQuery(
+  core::PropertyBuilder::createProperty("SQL select query")->isRequired(true)->withDescription(
+    "The SQL select query to execute. The query can be empty, a constant value, or built from attributes using Expression Language. "
+    "If this property is specified, it will be used regardless of the content of incoming flowfiles. "
+    "If this property is empty, the content of the incoming flow file is expected to contain a valid SQL select query, to be issued by the processor to the database. "
+    "Note that Expression Language is not evaluated for flow file contents.")->supportsExpressionLanguage(true)->build());
+
+const core::Property ExecuteSQL::s_maxRowsPerFlowFile(
+	core::PropertyBuilder::createProperty("Max Rows Per Flow File")->isRequired(true)->withDefaultValue<int>(0)->withDescription(
+		"The maximum number of result rows that will be included intoi a flow file. If zero then all will be placed into the flow file")->supportsExpressionLanguage(true)->build());
+
+const core::Relationship ExecuteSQL::s_success("success", "Successfully created FlowFile from SQL query result set.");
+
+static const std::string ResultRowCount = "executesql.row.count";
+
+ExecuteSQL::ExecuteSQL(const std::string& name, utils::Identifier uuid)
+  : SQLProcessor(name, uuid), max_rows_(0) {
+}
+
+ExecuteSQL::~ExecuteSQL() {
+}
+
+void ExecuteSQL::initialize() {
+  //! Set the supported properties
+  setSupportedProperties( { dbControllerService(), outputFormat(), s_sqlSelectQuery, s_maxRowsPerFlowFile});
+
+  //! Set the supported relationships
+  setSupportedRelationships( { s_success });
+}
+
+void ExecuteSQL::processOnSchedule(const core::ProcessContext &context) {
+  initOutputFormat(context);
+
+  context.getProperty(s_sqlSelectQuery.getName(), sqlSelectQuery_);
+  context.getProperty(s_maxRowsPerFlowFile.getName(), max_rows_);
+}
+
+void ExecuteSQL::processOnTrigger(core::ProcessSession &session) {
+  auto statement = connection_->prepareStatement(sqlSelectQuery_);
+
+  auto rowset = statement->execute();
+
+  int count = 0;
+  size_t rowCount = 0;
+  sql::JSONSQLWriter sqlWriter(isJSONPretty());
+  sql::SQLRowsetProcessor sqlRowsetProcessor(rowset, { &sqlWriter });
+
+  // Process rowset.
+  do {
+    rowCount = sqlRowsetProcessor.process(max_rows_ == 0 ? std::numeric_limits<size_t>::max() : max_rows_);
+    count++;
+    if (rowCount == 0)
+      break;
+
+    const auto& output = sqlWriter.toString();
```

Comment:
`SQLWriter::toString` returns an `std::string`, not `const std::string&`. Assigning this temporary to a const& will work because of reference lifetime extension, but it has no benefits, and it is misleading.

Reply:
It is an alternative to `move` copy constructor and can be used without knowing if returned object has it or not.

Reply:
In this case we know pretty well that `toString` returns a std::string and that it has a move constructor, not to mention in this case RVO would kick in.

Reply:
Changed in QueryDatabaseTable as well.

Reply:
Thanks!